BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java

```diff
@@ -105,33 +107,38 @@ private void removeTeleportedPlayers(){
 
 					@Override
 					public void run() {
-						List<Player> toRemove = new ArrayList<Player>();
-						for (Player p: arrivedPlayers){
-							if (!p.isOnline()) {
-								toRemove.add(p);
+						List<UUID> toRemove = new LinkedList<UUID>();
+						long currTime = System.currentTimeMillis();
+						for (Entry <UUID, Long> entry: arrivedPlayers.entrySet()){
+							Player p = Bukkit.getPlayer(entry.getKey());
+							if (p == null) {
 								continue;
```

Comment:
Keep in mind that players are only added to this list in the join event. Being in this list guarantees that they arrived and either are still standing in a portal or their timer isnt over yet. Keeping them is intended, because it means they wont be bounced back immediately once they login, but instead can walk out of the portal and only are sent back if they reenter it.